### PR TITLE
Feature/settings validation

### LINKED ISSUE #51: Constraint validation
There should be a way to reject semantically invalid values (instead of just syntactically invalid as now).
This is setting-dependent.

* `f32` bounds (max, min)
* Rounding
* Complicated multi-settings constraints (iir filter coefficients leading to undesired zero/pole positions due to rounding)

### PR DESCRIPTION
**Pull request by ryan-summers**
This PR fixes #51 by adding support for a custom handler to be called whenever settings are updated.

Note: Currently, if a new setting is rejected, it still persists in the cached settings.

### CODE REVIEW

File: src/mqtt_client/mqtt_client.rs
Comment (ryan-summers) [inline src/mqtt_client/mqtt_client.rs:341]: I'm not super happy with this implementation. It makes us serialize the setting that is being updated on every change. We could alternatively just require that `Settings: Copy` and make a deep copy of the settings structure. It's unknown which one would be the least overhead.

Additionally, the code flow here is pretty convoluted and could likely be broken out into a separate function or refactored.

File: src/mqtt_client/mqtt_client.rs
Comment (ryan-summers) [inline src/mqtt_client/mqtt_client.rs:352]: Need to set updated to false if we rejected the settings.

I think we also need to republish the correct value.

File: tests/validation_failure.rs
Comment (ryan-summers) [inline tests/validation_failure.rs:105]: Check that `settings.error` has properly been de-asserted.

**Comment by ryan-summers** [comment]
@jordens Would you mind taking a look at the implementation here for some thoughts? I'm trying to answer the following questions:
1. Should settings requests be rejectable? 
2. If so, what should miniconf do on rejection? (e.g. restore settings, republish previous values, more?)

File: src/mqtt_client/mqtt_client.rs
Comment (jordens) [reply src/mqtt_client/mqtt_client.rs:341]: Hmm. Yes that doesn't look good.
1. To separate requested settings from actual active setting on mqtt we may want to separate those trees on the topic level.
2. I would usually expect the sequence to be (1) deser of the payload, (2) validation (3) application/update. Maybe 2 and 3 can both be done by the handler.

File: src/mqtt_client/mqtt_client.rs
Comment (jordens) [inline src/mqtt_client/mqtt_client.rs:280]: Ah. It took me five attempts to understand what "handled_update()" means.
Maybe just unify and have `update<F, E, T>(&mut self, handler: Option<F>)`?

File: src/mqtt_client/mod.rs
Comment (jordens) [inline src/mqtt_client/mod.rs:19]: I thought about it a bit more. I think we can skip this. Instead have the `Result::Ok` from the handler be returned from `update() -> Result` to the application. Then the application can decide based on that result if and what to do. We just need to make sure we expose the republication API to the application.
What do you think?

Comment (jordens) [reply src/mqtt_client/mod.rs:19]: And we should also give the handler two more bits of information if possible: a ref to the old settings and the path of the update.

File: src/mqtt_client/mqtt_client.rs
Comment (ryan-summers) [reply src/mqtt_client/mqtt_client.rs:280]: My onle goal with adding `handled_update()` was that it was a way to maintain backwards compatibility with the previous API.

The proposed API is really unergonomic if you don't have a handler, since Rust will complain that the type `F` cannot be automatically determined.

File: src/mqtt_client/mod.rs
Comment (ryan-summers) [reply src/mqtt_client/mod.rs]: I've added additional info to the handler.

However, we can't allow the `update()` to return the result of the handler, since the handler may be called N times due to the MQTT poll loop. We can refactor that in the future, but for now, I've updated it so that the handler has to return Result<(), E>`

File: src/mqtt_client/mqtt_client.rs
Comment (jordens) [inline src/mqtt_client/mqtt_client.rs:54]: I would not reset the timeout on incoming updates. That creates unnecessary dependencies.

File: src/mqtt_client/mqtt_client.rs
Comment (jordens) [inline src/mqtt_client/mqtt_client.rs:268]:
```suggestion
    /// Update the MQTT interface and service the network. Pass any settings changes to the handler supplied.
```
```suggestion
    /// Update the MQTT interface and service the network. Handle any settings changes.
```

File: src/mqtt_client/mqtt_client.rs
Comment (jordens) [inline src/mqtt_client/mqtt_client.rs:394]: I would give the handler the task to do the settings copy. That would support more complicated use cases where the desired new settings are more easily computed from the old and not the new and save one of the two settings copying.

```suggestion
        self.handled_update(|_path, old, new| { *old = new; Result::<(), ()>::Ok(()) })
```

File: src/mqtt_client/mod.rs
Comment (jordens) [reply src/mqtt_client/mod.rs]: Can't we just exit the loop on error? The unhandled incoming data will still be there (and presumably be handled next time) and the TX buffer is guaranteed to drain (so that there will be no deadlock).

File: src/mqtt_client/mqtt_client.rs
Comment (jordens) [reply src/mqtt_client/mqtt_client.rs]: Ack. Then maybe worth thinking a bit more about the docstring to make it clear what the function is and does. I think that would help me already.

File: src/mqtt_client/mod.rs
Comment (ryan-summers) [reply src/mqtt_client/mod.rs]: As of right now, minimq doesn't provide a mechanism to return an `Error` from inside the closure - the result of processing is unused when considering continuing iteration:
https://github.com/quartiq/minimq/blob/e8a072726ca94efbd2048917218cb2ba24b5c599/src/mqtt_client.rs#L413

I'm not opposed to refactoring minimq to support this behavior - I'll open up an issue.

Comment (jordens) [reply src/mqtt_client/mod.rs]: Ack. I'll leave it to you whether you want to wait for that in `minimq` and then follow up here or land this here first and then refactor later. But it'll ripple through the stack.

Comment (ryan-summers) [reply src/mqtt_client/mod.rs]: Actually, after testing it out more, the application can just pass information back via accessing external state in the closure, so it can track any results that way if need be. I actually think the `Result` of this closure should be used for transferring a response back to the original requestee.

For example:
```rust
let mut republish_required = false;
client.handled_update(|path, old, new| {
    if new.threshold > MAX_THRESHOLD {
        return Err("Requested threshold above maximum";
    }

    *old = new.clone();
    Ok(())
}).unwrap()
```

File: src/mqtt_client/mqtt_client.rs
Comment (ryan-summers) [reply src/mqtt_client/mqtt_client.rs]: I added a complete code example to show the user.

**Review by jordens — APPROVED** [review]

**Pull request merged**
